File: 12/pcMonitor.java
import java.util.*;
import java.lang.*;
import java.util.concurrent.locks.*;

    public class pcMonitor {

    	//private Lock cerrojo= new ReentrantLock();
    	private final Condition no_vacio = cerrojo.newCondition();
    	private final Condition no_lleno = cerrojo.newCondition();
        int numSlots;
        double[]vector;
        Random rd = new Random();
        int putIn,takeOut;
        int cont;

      public pcMonitor(int numSlots){
        this.numSlots=numSlots;
        vector=new double[numSlots];
      }

      public synchronized void insertar () {
      	 // cerrojo.lock();
      	  //try{
        while (cont == numSlots)
          try {
          	  no_vacio.await();
          } catch (InterruptedException e) {}
        vector[putIn] =(Math.pow(rd.nextDouble(),2)+Math.pow(rd.nextDouble(),2));
        putIn = (putIn + 1) % numSlots;
        cont++;
       no_lleno.signalAll();
       	//  }finally{cerrojo.unlock();}
      														}

      public synchronized double extraer () {
      	// cerrojo.lock();
      	 //try{
        double valor;
        while (cont == 0)
          try {
            no_lleno.await();
          } catch (InterruptedException e) {
            System.err.println("wait interrumpido");
          }
        valor = vector[takeOut];
        takeOut = (takeOut + 1) % numSlots;
        cont--;
        no_vacio.signalAll();
        return valor;
        // }finally{cerrojo.unlock();}
      }
    }//Buffer

File: 12/pcMonitor.cpp
#include <iostream>
#include <condition_variable>
#include <mutex>
using namespace std;

struct pcMonitor{//EN java es el monitor cadena
	int numSlots=0,putIn=0,takeOut=0,cont=0;
	double* buffer;
	std::condition_variable not_full;
	std::condition_variable not_empty;
	std::mutex lock;


	pcMonitor(int nslots){
		numSlots=nslots;
		buffer= new double[nslots];
	}

	void producir(double valor){
		std::unique_lock<std::mutex> l(lock);
		not_full.wait(l,[this](){return cont!=numSlots;});//se realiza el wair cuando el{} es falso

		buffer[putIn]=valor;
		putIn=(putIn+1)%numSlots;
		++cont;
		not_empty.notify_one();

	}

	double consumir(){
		std::unique_lock<std::mutex> l(lock);
		not_empty.wait(l,[this](){return cont!=0;});

		double res=buffer[takeOut];
		takeOut=(takeOut+1)%numSlots;
		--cont;
		not_full.notify_one();
		return res;
	}
};
